distributed/tests/test_semaphore.py

```diff
@@ -0,0 +1,156 @@
+import pickle
+from time import sleep
+
+from distributed import Client, Semaphore, get_client
+from distributed.metrics import time
+from distributed.utils_test import cluster, gen_cluster  # noqa F401
+
+import dask
+from dask.distributed import Client
+
+
+@gen_cluster(client=True)
+async def test_semaphore(c, s, a, b):
+    semaphore = await Semaphore(max_leases=2, name="exasol_db")
+
+    result = await semaphore.acquire()  # allowed_leases: 2 - 1 -> 1
+    assert result is True
+
+    second = await semaphore.acquire()  # allowed_leases: 1 - 1 -> 0
+    assert second is True
+    start = time()
+    result = await semaphore.acquire(timeout=0.025)  # allowed_leases: 0 -> False
+    stop = time()
+    assert stop - start < 0.2
+    assert result is False
+
+
+@gen_cluster(client=True)
+async def test_serializable(c, s, a, b):
+    sem = await Semaphore(max_leases=2, name="x")
+    res = await sem.acquire()
+    assert len(s.extensions["semaphores"].leases["x"]) == 1
+    assert res
+    sem2 = pickle.loads(pickle.dumps(sem))
+    assert sem2.name == sem.name
+    assert sem2.client.scheduler.address == sem.client.scheduler.address
+
+    # actual leases didn't change
+    assert len(s.extensions["semaphores"].leases["x"]) == 1
+
+    res = await sem2.acquire()
+    assert res
+
+    # Ensure that both objects access the same semaphore
+    res = await sem.acquire(timeout=0.025)
+
+    assert not res
+    res = await sem2.acquire(timeout=0.025)
+
+    assert not res
+
+
+@gen_cluster(client=True)
+async def test_release_simple(c, s, a, b):
+    def f(x, semaphore):
+        with semaphore:
+            assert semaphore.name == "x"
+            return x + 1
+
+    sem = await Semaphore(max_leases=2, name="x")
+    futures = c.map(f, list(range(10)), semaphore=sem)
+    await c.gather(futures)
+
+
+@gen_cluster(client=True)
+async def test_acquires_with_timeout(c, s, a, b):
+    sem = await Semaphore(1, "x")
+    assert await sem.acquire(timeout=0.025)
+    assert not await sem.acquire(timeout=0.025)
+    await sem.release()
+    assert await sem.acquire(timeout=0.025)
+    await sem.release()
+
+
+def test_timeout_sync():
```

Comment:
```suggestion
def test_timeout_sync(c, s, a, b):
```

I believe that there are pytest fixtures that now replace the cluster/client manual construction.

Reply:
```
  def test_timeout_sync(c, s, a, b):
E       fixture 'c' not found
```

Reply:
Might be that you need to import it first. With fixtures defined in a proper module, the rules are slightly different compared to `conftest` defined fixtures

Reply:
ah my bad

Reply:
This is in test_client.py

```python
from distributed.utils_test import (  # noqa: F401
    client as c,
    s,
    a,
    b,
)
```